### PR TITLE
Add a new command to inspect redundant dependencies

### PR DESCRIPTION
**Pull request by rofle100lvl**
### Short description 📝

A week ago we added Implicit Imports finder functional in tuist. But now I came up with second step, which costs almost nothing and provide very helpful feature for the user. Redundant Dependencies. If user declared dependency in tuist manifest and dont use it we might want to provide that information. This pr is about that. 

How it works? 

Absolutely, the same with Implicit Imports. But with one difference. 

var imports - imports found in sources of target
var dependencies - modules which are current target depends on.

In implicit imports finder we make `imports subtracting dependencies` and find all dependencies which are imported, but not included in tuist manifest.
But if we de in opposite site `dependencies subtracting imports`. We get all dependencies which are declared in tuist manifest but not used in code.

For Tuist it works like t[hat](https://pastebin.com/8ZvgQ0Nr) 
### How to test the changes locally 🧐

> Include a set of steps for the reviewer to test the changes locally (see [the documentation](https://docs.tuist.io/contributors/get-started) for reference).

### Contributor checklist ✅

- [ ] The code has been linted using run `mise run lint-fix`
- [ ] The change is tested via unit testing or acceptance testing, or both
- [ ] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Review by kwridan — COMMENTED** [review]
Thanks @rofle100lvl 

This is a very nice addition 👌

Redundant imports might be a bit more nuanced than implicit imports 😅, one use case I can think of that can flag false positives is resource bundle targets. 

e.g.

Running against the following fixture:

https://github.com/tuist/tuist/tree/main/fixtures/ios_app_with_framework_and_resources

```sh
 $ tuist inspect redundant-imports 
Loading and constructing the graph
It might take a while if the cache is empty
The following redundant dependencies were found:
 - StaticFramework5 redundantly depends on: StaticFramework5_StaticFramework5
 - App redundantly depends on: StaticFrameworkResources
 - StaticFramework4 redundantly depends on: StaticFramework4_StaticFramework4
 - StaticFramework2 redundantly depends on: StaticFramework2Resources
 - StaticFramework3 redundantly depends on: StaticFramework3_StaticFramework3
```

The identified targets will not contain any code and as such won't have any corresponding imports by modules that depend on them. Some are auto-generated by Tuist itself (e.g. the modules repeated names with `_` -`StaticFramework3_StaticFramework3`)

Perhaps redundant imports can be scoped to `.framework`, `.staticFramework`, `.staticLibrary`, `.dynamicLibrary` targets?

**Comment by rofle100lvl** [comment]
> Perhaps redundant imports can be scoped to `.framework`, `.staticFramework`, `.staticLibrary`, `.dynamicLibrary` targets?
Thank you so much for that corner case @kwridan!
Hm, it looks like for that case it's required to exclude .bundle target it works. Now we see
`It might take a while if the cache is empty
We did not find any redundant dependencies in your project.
`
Can you please think about other corner cases may appear?

**Comment by sphanley** [comment]
A couple of corner cases I'm seeing: 
* App extensions are flagged as redundant
* When a test target references its host app, that's flagged as redundant

**Comment by rofle100lvl** [comment]
@kwridan Thanks! After small research i think you are right. Set up filter on` .framework`, `.staticFramework`, `.staticLibrary`, `.dynamicLibrary` types

**Review by danieleformichelli — APPROVED** [review]

**Review by kwridan — DISMISSED** [review]
thanks @rofle100lvl !

One additional edge case I discovered is in ObjC test targets, they can reference headers from the target they are testing without the module / target prefix 🙈

```objc
// ObjCModuleTests

#import "SomeHeader.h" // From ObjCModule

// ....
```

This can occur if the header being imported isn't a public header of the Objective C module as such wouldn't be imported via `#import <ObjCModule/SomeHeader.h>`. It's an edge case that can occur if no public headers are ever referenced in the test target.

File: Sources/TuistKit/Commands/InspectCommand.swift [hunk 1]
```diff
@@ -4,7 +4,10 @@ struct InspectCommand: AsyncParsableCommand {
     static var configuration: CommandConfiguration {
         CommandConfiguration(
             commandName: "inspect",
```
Comment (fortmarek) [inline Sources/TuistKit/Commands/InspectCommand.swift:6]: Let's add abstract to this command as well.

The entry is missing in `tuist --help`, which looks out of place:
image

File: Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift [hunk 2]
```diff
@@ -31,7 +52,12 @@ enum InspectImplicitImportsServiceError: FatalError, Equatable {
     }
 }
 
-final class InspectImplicitImportsService {
+enum InspectType {
+    case redundant
+    case implicit
+}
+
+final class InspectImportsService {
```
Comment (fortmarek) [inline Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift:60]: I'd consider keeping a service per command and use abstractions for the linting process itself.

File: Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift [hunk 3]
```diff
@@ -46,41 +72,74 @@ final class InspectImplicitImportsService {
...
         let graph = try await generator.load(path: path)
-        let issues = try await lint(graphTraverser: GraphTraverser(graph: graph))
+        let issues = try await lint(graphTraverser: GraphTraverser(graph: graph), inspectType: inspectType)
         guard issues.isEmpty else {
-            throw InspectImplicitImportsServiceError.implicitImportsFound(issues)
+            switch inspectType {
+            case .redundant:
+                throw InspectRedundantImportsServiceError.redundantImportsFound(issues)
+            case .implicit:
+                throw InspectImplicitImportsServiceError.implicitImportsFound(issues)
```
Comment (fortmarek) [inline Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift:89]: These errors suggest there are two different services for inspecting imports. That is not true – we currently have only one.

File: Tests/TuistKitTests/Services/Inspect/LintRedundantImportsServiceTests.swift [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+@testable import TuistKit
+
+final class LintRedundantImportsServiceTests: TuistUnitTestCase {
+    private var configLoader: MockConfigLoading!
+    private var generatorFactory: MockGeneratorFactorying!
+    private var targetScanner: MockTargetImportsScanning!
+    private var subject: InspectImportsService!
+    private var generator: MockGenerating!
+
+    override func setUp() async throws {
```
Comment (fortmarek) [inline Tests/TuistKitTests/Services/Inspect/LintRedundantImportsServiceTests.swift:22]:
```suggestion
    override func setUp() {
```

No need to use `async throws` afaict

File: Tests/TuistKitTests/Services/Inspect/LintRedundantImportsServiceTests.swift [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        targetScanner = MockTargetImportsScanning()
+        generator = MockGenerating()
+        subject = InspectImportsService(
+            generatorFactory: generatorFactory,
+            configLoader: configLoader,
+            targetScanner: targetScanner
+        )
+    }
+
+    override func tearDown() async throws {
```
Comment (fortmarek) [inline Tests/TuistKitTests/Services/Inspect/LintRedundantImportsServiceTests.swift:35]:
```suggestion
    override func tearDown() {
```

File: Tests/TuistKitTests/Services/Inspect/LintRedundantImportsServiceTests.swift [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        given(generator).load(path: .value(path)).willReturn(graph)
+        given(targetScanner).imports(for: .value(app)).willReturn(Set([]))
+        given(targetScanner).imports(for: .value(framework)).willReturn(Set([]))
+
+        let expectedError = InspectRedundantImportsServiceError.redundantImportsFound([
+            InspectImportsServiceErrorIssue(target: "App", implicitDependencies: Set(["Framework"])),
+        ])
+
+        // When
+        await XCTAssertThrowsSpecific({ try await subject.run(path: path.pathString, inspectType: .redundant) }, expectedError)
```
Comment (fortmarek) [inline Tests/TuistKitTests/Services/Inspect/LintRedundantImportsServiceTests.swift:68]:
```suggestion
        await XCTAssertThrowsSpecific(try await subject.run(path: path.pathString, inspectType: .redundant), expectedError)
```

File: Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift [hunk 7]
```diff
@@ -46,41 +72,74 @@ final class InspectImplicitImportsService {
...
+            "We did not find any \(inspectType == .implicit ? "implicit" : "redundant") dependencies in your project."
+        )
     }
 
-    private func lint(graphTraverser: GraphTraverser) async throws -> [InspectImplicitImportsServiceErrorIssue] {
-        let allTargets = graphTraverser
+    private func lint(
+        graphTraverser: GraphTraverser,
+        inspectType: InspectType
+    ) async throws -> [InspectImportsServiceErrorIssue] {
```
Comment (fortmarek) [inline Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift:101]: I'd move this to a service its own that would be shared between `InspectImplicitImportsService` and `InspectRedundantImportsService`. For example, it can be called `GraphImportsLinter`.

Additionally, I'd return `LintingIssue` instead of a custom issue that is then transformed into an error. We don't need to follow the `GraphLinting` protocol 1-1, but we should be _very_ close to it. They both solve a very similar problem, so there's no need to diverge here.

File: Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift
Comment (fortmarek) [inline Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift:1]: The file no longer reflects the actual name of the file. Also, it has a typo (`Inspec` vs `Inspect`)

File: Tests/TuistKitTests/Services/Inspect/InspectImplicitImportsServiceTests.swift [hunk 8]
```diff
@@ -57,11 +57,11 @@ final class LintImplicitImportsServiceTests: TuistUnitTestCase {
         given(targetScanner).imports(for: .value(framework)).willReturn(Set([]))
 
         let expectedError = InspectImplicitImportsServiceError.implicitImportsFound([
-            InspectImplicitImportsServiceErrorIssue(target: "App", implicitDependencies: Set(["Framework"])),
+            InspectImportsServiceErrorIssue(target: "App", implicitDependencies: Set(["Framework"])),
         ])
 
         // When
-        await XCTAssertThrowsSpecific({ try await subject.run(path: path.pathString) }, expectedError)
+        await XCTAssertThrowsSpecific({ try await subject.run(path: path.pathString, inspectType: .implicit) }, expectedError)
```
Comment (fortmarek) [inline Tests/TuistKitTests/Services/Inspect/InspectImplicitImportsServiceTests.swift:64]:
```suggestion
        await XCTAssertThrowsSpecific(try await subject.run(path: path.pathString, inspectType: .implicit), expectedError)
```

File: Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift [hunk 9]
```diff
@@ -46,41 +72,74 @@ final class InspectImplicitImportsService {
...
             .allInternalTargets()
 
+        if inspectType == .redundant {
+            allTargets = allTargets.filter {
+                switch $0.target.product {
+                case .staticLibrary, .staticFramework, .dynamicLibrary, .framework: true
+                default: false
+                }
+            }
+        }
```
Comment (fortmarek) [inline Sources/TuistKit/Services/Inspect/InspecImplicitImportsService.swift:112]:
```suggestion
        let allTargets = graphTraverser
            .allInternalTargets()
            .filter {
              switch inspectType {
                case .redundant:
                  return switch $0.target.product { ... }
                case .implicit:
                  return true
              }
            }
```

Let's try to use exhaustive switches where we can. Also, no need to make `allTargets` a variable.

**Review by fortmarek — CHANGES_REQUESTED** [review]
Thanks a lot!

This will be a great addition 🙂

**Comment by fortmarek** [comment]
@rofle100lvl any chance you'll have a time to look at this? There's definitely interest for this feature 😉 https://community.tuist.io/t/unused-import-detection/195

**Comment by rofle100lvl** [comment]
@fortmarek It's cool that somebody get same ideas as me! I will look into this, but firstly, I want to come up with decision about SPM recursive checking. I think that it has more priority right now.

File: Sources/TuistKit/Services/Inspect/GraphImportsLinter.swift [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    private let targetScanner: TargetImportsScanning
+
+    init(targetScanner: TargetImportsScanning = TargetImportsScanner()) {
+        self.targetScanner = targetScanner
+    }
+
+    func lint(
+        graphTraverser: GraphTraverser,
+        inspectType: InspectType
+    ) async throws -> [LintingIssue] {
```
Comment (fortmarek) [inline Sources/TuistKit/Services/Inspect/GraphImportsLinter.swift:20]: I'd consider splitting the method and its implementation into multiple ones to make the code easier to follow.

I'd envision something like:
```swift
func lintImplicitImports(graphTraverser: GraphTraversing) {
  let targetsToLint = graphTraverser.allInternalTargets().union(graphTraverser.allExternalTargets()).filter { ... }
  return targetImportsMap.flatMap { (target, dependencyImports) in
   let explicitTargetDependencies = explicitTargetDepencies(target)
    dependencyImports.map { dependencyImport in
      if !explicitTargetDependencies.contains(where: { $0.name == dependencyImport.name }) {
        return LintingIssue(...)
     }
    }
  }
}

struct DependencyImport {
  let name: String
  let count: Int
}

/// - Returns: A map of targets with dependency imports
private func targetImportsMap(graphTraverser: GraphTraversing, targetsToLint: [Target]) -> [Target: [DependencyImport]] {
  ...
}
```

Comment (fortmarek) [reply Sources/TuistKit/Services/Inspect/GraphImportsLinter.swift:27]: @rofle100lvl this doesn't seem to have been addressed. Do you think it's not the right approach?

**Comment by fortmarek** [comment]
@rofle100lvl would be great to move forward on this, but would like to resolve this discussion first: https://github.com/tuist/tuist/pull/6698#discussion_r1888703339

Can you let me know your thoughts? 🙂

**Comment by rofle100lvl** [comment]
> @rofle100lvl would be great to move forward on this, but would like to resolve this discussion first: [#6698 (comment)](https://github.com/tuist/tuist/pull/6698#discussion_r1888703339)
> 
> Can you let me know your thoughts? 🙂
Marek, I made method separation you wanted to do. Below lintImplicitImports method you can see targetImportsMap method which should be exactly what you wrote about

**Review by fortmarek — APPROVED** [review]
What do you think about adding acceptance tests? Otherwise, this looks good to go 🙏 

Sorry again for the slow turnaround time. I'll try to be better at this.

**Review by fortmarek — APPROVED** [review]

**Comment by fortmarek** [comment]
Thanks @rofle100lvl for this amazing contribution 👏

**Comment by adarhef** [comment]
Hey, thank you for this! Love to finally get some tooling around unused dependencies.

I'm having an issue with this that I'd like to get some clarification on.

I have a target that depended on TCA. I noticed the imports in the target and saw that I was only importing `IdentifiedCollections`. So I reduced the scope to only depend on `IdentifiedCollections`. Then I still got an error that I'm redundantly depending on `InternalCollectionsUtilities, OrderedCollections`.
These are modules from a downstream dependency of `swift-identified-collections`, which is Apple's `swift-collections`.
At this point logging these modules as redundant is just noise because it's not actionable. The _only_ product of `swift-identified-collections` is `IdentifiedCollections` and that's the only product I need to be depending on, as it contains the types I want to use.

It looks like to me that we should be a little more clever in what we flag to make sure we're not alerting about supposedly-redundant downstream dependencies. 
There's already valuable insights in the output, but the signal-to-noise ratio on it is not good enough for me to try to dig into it as it stands. I'd really like to get value from it.
